### PR TITLE
Start adding test for listeners

### PR DESCRIPTION
**Pull request by chrisgilmerproj**
## Description

This begins the process of adding better testing around our server listeners.

## Reviewer Notes

Is there anything you would like reviewers to give additional scrutiny?

## Setup

Add any steps or code to run in this section to help others prepare to run your code:

```sh
echo "Code goes here"
```

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/tree/master/docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/tree/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/tree/master/docs/database.md#zero-downtime-migrations))
  * [ ] Have been communicated to #dp3-engineering
  * [ ] Secure migrations have been tested using `scripts/run-prod-migrations`
* [ ] There are no aXe warnings for UI.
* [ ] This works in [Supported Browsers and their phone views](https://github.com/transcom/mymove/tree/master/docs/adr/0016-Browser-Support.md) (Chrome, Firefox, IE, Edge).
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] User facing changes have been reviewed by design.
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](tbd) for this change
* [this article](tbd) explains more about the approach used.

## Screenshots

If this PR makes visible UI changes, an image of the finished UI can help reviewers and casual
observers understand the context of the changes. A before image is optional and
can be included at the submitter's discretion.

Consider using an animated image to show an entire workflow instead of using multiple images. You may want to use GIPHY CAPTURE for this! 📸

_Please frame screenshots to show enough useful context but also highlight the affected regions._

### CODE REVIEW

**Comment by pjdufour-truss** [comment]
Can you dump the TLS information from the request within the handler to confirm in some way that it did a mutual TLS handshake?

File: pkg/server/server_test.go [hunk 1]
```diff
@@ -186,3 +187,78 @@ func (suite *serverSuite) TestHTTPServerConfig() {
...
+		// Check the TLS connection from inside the request
+		connState := r.TLS
+		suite.Equal(tls.VersionTLS12, int(connState.Version))
+		suite.True(connState.HandshakeComplete)
+		suite.False(connState.DidResume)
+		suite.Equal("", connState.NegotiatedProtocol)
+		suite.True(connState.NegotiatedProtocolIsMutual)
+		suite.Equal("localhost", connState.ServerName)
+		suite.Equal("Snake Oil", connState.PeerCertificates[0].Subject.Organization[0])
+		suite.Equal("Snake Oil", connState.VerifiedChains[0][0].Subject.Organization[0])
```
Comment (chrisgilmerproj) [inline pkg/server/server_test.go]: @pjdufour-truss - here's where I'm testing the TLS connection inside the handler. What do you think? Is this a good test?

**Comment by chrisgilmerproj** [comment]
Need a test that verifies that a request is blocked if the certificate or cert chain is invalid.

File: pkg/server/server_test.go [hunk 2]
```diff
@@ -186,3 +187,78 @@ func (suite *serverSuite) TestHTTPServerConfig() {
 	suite.Equal(httpsServer.Addr, "127.0.0.1:8080")
 	suite.Equal(suite.httpHandler, httpsServer.Handler)
 }
+
+func (suite *serverSuite) TestTLSConfigWithRequest() {
```
Comment (chrisgilmerproj) [inline pkg/server/server_test.go]: Need a test that fails connection and never reaches the handler.

Comment (chrisgilmerproj) [reply pkg/server/server_test.go]: Test certs are here: https://github.com/transcom/mymove/tree/master/pkg/server/testdata

File: cmd/milmove/serve.go [hunk 3]
```diff
@@ -286,6 +287,44 @@ func indexHandler(buildDir string, logger logger) http.HandlerFunc {
...
+		Port:         port,
+		Logger:       logger,
+		HTTPHandler:  httpHandler,
+		ClientAuth:   tls.NoClientCert,
+		Certificates: certificates,
+	})
+}
+
+// CreateNamedMutualTLSServer returns a named server with mutual TLS configured
+func CreateNamedMutualTLSServer(host string, port int, logger server.Logger, httpHandler http.Handler, certificates []tls.Certificate, clientCAs *x509.CertPool) (*server.NamedServer, error) {
```
Comment (chrisgilmerproj) [inline cmd/milmove/serve.go]: Test this function in this file: https://github.com/transcom/mymove/blob/master/cmd/milmove/main_test.go

**Comment by mr337** [comment]
@chrisgilmerproj Quick review if you don't mind.

I was trying to build a new cert from the existing CA but missing a key to sign a new cert to test using a valid but expired cert.

**Comment by chrisgilmerproj** [comment]
🚀 - I approve the stuff from @mr337

**Pull request merged**